src/librustc_metadata/decoder.rs

```diff
@@ -1180,14 +1141,13 @@ fn struct_field_family_to_visibility(family: Family) -> ty::Visibility {
     }
 }
 
-pub fn get_struct_field_names(intr: &IdentInterner, cdata: Cmd, id: DefIndex)
-    -> Vec<ast::Name> {
+pub fn get_struct_field_names(cdata: Cmd, id: DefIndex) -> Vec<ast::Name> {
     let item = cdata.lookup_item(id);
     let mut index = 0;
     reader::tagged_docs(item, tag_item_field).map(|an_item| {
-        item_name(intr, an_item)
+        item_name(an_item)
     }).chain(reader::tagged_docs(item, tag_item_unnamed_field).map(|_| {
-        let name = intr.intern(&index.to_string());
+        let name = token::with_ident_interner(|interner| interner.intern(index.to_string()));
```

Comment:
Yeah, I'd like `token::intern` to be generic but it's a syntax-[breaking-change] since `token::intern(&my_string)` where `my_string: String` wouldn't typecheck (`intern(my_string)` or `intern(&*my_string)` would, though).

I was planning on making `token::intern` generic in a future PR with other interner-related syntax-[breaking-change]s.

Also, I agree that `with_ident_interner` should eventually be private and that `token::intern` should eventually be `Symbol::from` (or `Symbol::intern` to contrast with `Symbol::gensym`).